File: solutions/elixir/complex-numbers/lib/complex_numbers.ex
defmodule ComplexNumbers do
  import Kernel, except: [div: 2]

  @typedoc """
  In this module, complex numbers are represented as a tuple-pair containing the real and
  imaginary parts.
  For example, the real number `1` is `{1, 0}`, the imaginary number `i` is `{0, 1}` and
  the complex number `4+3i` is `{4, 3}'.
  """
  @type complex :: {float, float}

  @doc """
  Return the real part of a complex number
  """
  @spec real(z :: complex) :: float
  def real({a, _}), do: a

  @doc """
  Return the imaginary part of a complex number
  """
  @spec imaginary(z :: complex) :: float
  def imaginary({_, b}), do: b

  @doc """
  Multiply two complex numbers, or a real and a complex number
  """
  @spec mul(z :: complex | float, y :: complex | float) :: complex
  def mul(x, {c, d}) when is_number(x), do: mul({x, 0}, {c, d})
  def mul({a, b}, y) when is_number(y), do: mul({a, b}, {y, 0})

  def mul({a, b}, {c, d}) do
    {a * c - b * d, b * c + a * d}
  end

  @doc """
  Add two complex numbers, or a real and a complex number
  """
  @spec add(z :: complex | float, y :: complex | float) :: complex
  def add(x, {c, d}) when is_number(x), do: add({x, 0}, {c, d})
  def add({a, b}, y) when is_number(y), do: add({a, b}, {y, 0})

  def add({a, b}, {c, d}) do
    {a + c, b + d}
  end

  @doc """
  Subtract two complex numbers, or a real and a complex number
  """
  @spec sub(z :: complex | float, y :: complex | float) :: complex
  def sub(x, {c, d}) when is_number(x), do: sub({x, 0}, {c, d})
  def sub({a, b}, y) when is_number(y), do: sub({a, b}, {y, 0})

  def sub({a, b}, {c, d}) do
    {a - c, b - d}
  end

  @doc """
  Divide two complex numbers, or a real and a complex number
  """
  @spec div(z :: complex | float, y :: complex | float) :: complex
  def div(x, {c, d}) when is_number(x), do: div({x, 0}, {c, d})
  def div({a, b}, y) when is_number(y), do: div({a, b}, {y, 0})

  def div({a, b}, {c, d}) do
    {(a * c + b * d) / (c * c + d * d), (b * c - a * d) / (c * c + d * d)}
  end

  @doc """
  Absolute value of a complex number
  """
  @spec abs(z :: complex) :: float
  def abs({a, b}) do
    :math.sqrt(a * a + b * b)
  end

  @doc """
  Conjugate of a complex number
  """
  @spec conjugate(z :: complex) :: complex
  def conjugate({a, b}) do
    {a, -b}
  end

  @doc """
  Exponential of a complex number
  """
  @spec exp(z :: complex) :: complex
  def exp({a, b}) do
    {:math.exp(a) * :math.cos(b), :math.exp(a) * :math.sin(b)}
  end
end

File: solutions/prolog/complex-numbers/complex_numbers.pl
% A complex number is a number in the form a + b * i
% where a and b are real and i satisfies i^2 = -1.
real((A, _B), R) :- R = A.
imaginary((_A, B), I) :- I = B.

% conjugate of a + b * i is a - b * i
conjugate(C, Conjugate) :-
    real(C, A), imaginary(C, B),
    real(Conjugate, A), imaginary(Conjugate, 0-B).

% absolute value of complex a + b * i is real sqrt(a^2 + b^2)
abs(C1, Abs) :-
    real(C1, A), imaginary(C1, B),
    Abs is sqrt(A*A+B*B).

% Addition (a + i * b) + (c + i * d) = (a + c) + (b + d) * i
add(C1, C2, CAdd) :-
    real(C1, A), imaginary(C1, B),
    real(C2, C), imaginary(C2, D),
    real(CAdd, A+C), imaginary(CAdd, B+D).

% Subtraction (a + i * b) - (c + i * d) = (a - c) + (b - d) * i.
sub(C1, C2, CSub) :-    
    real(C1, A), imaginary(C1, B),
    real(C2, C), imaginary(C2, D),
    real(CSub, A-C), imaginary(CSub, B-D).

% Multiplication (a + i * b) * (c + i * d) 
%   = (a * c - b * d) + (b * c + a * d) * i
mul(C1, C2, CMul) :-
    real(C1, A), imaginary(C1, B),
    real(C2, C), imaginary(C2, D),
    real(CMul, A*C-B*D), imaginary(CMul, B*C+A*D).

% Division (a + i * b) / (c + i * d) 
%   = (a * c + b * d)/(c^2 + d^2) + (b * c - a * d)/(c^2 + d^2) * i
div(C1, C2, CDiv) :-     
    real(C1, A), imaginary(C1, B),
    real(C2, C), imaginary(C2, D),
    real(CDiv, (A*C+B*D)/(C*C+D*D)),
    imaginary(CDiv, (B*C-A*D)/(C*C+D*D)).
